openflow15/match.go

```diff
@@ -0,0 +1,2173 @@
+	}
+	m.FLabel = binary.BigEndian.Uint32(data[0:])
+	return nil
+}
+
+func NewIpv6FLabelField(flabel uint32, flabelMask *uint32) *MatchField {
+	f := new(MatchField)
+	f.Class = OXM_CLASS_OPENFLOW_BASIC
+	f.Field = OXM_FIELD_IPV6_FLABEL
+	f.HasMask = true
+
+	flabelField := new(Ipv6FLabelField)
+	flabelField.FLabel = flabel
+	f.Value = flabelField
+	f.Length = uint8(flabelField.Len())
+
+	// Add the mask
+	if flabelMask != nil {
+		mask := new(Ipv6FLabelField)
+		mask.FLabel = *flabelMask
+		f.Mask = mask
+		f.HasMask = true
+		f.Length += uint8(mask.Len())
+	}
+	return f
+}
+
+// IPV6_DST field
+type Ipv6DstField struct {
+	Ipv6Dst net.IP
+}
+
+func (m *Ipv6DstField) Len() uint16 {
+	return 16
+}
+func (m *Ipv6DstField) MarshalBinary() (data []byte, err error) {
+	data = make([]byte, 16)
+	copy(data, m.Ipv6Dst)
+	return
+}
+
+func (m *Ipv6DstField) UnmarshalBinary(data []byte) error {
+	m.Ipv6Dst = make([]byte, 16)
+	copy(m.Ipv6Dst, data)
+	return nil
+}
+
+// Return a MatchField for ipv6 dest addr
+func NewIpv6DstField(ipDst net.IP, ipDstMask *net.IP) *MatchField {
+	f := new(MatchField)
+	f.Class = OXM_CLASS_OPENFLOW_BASIC
+	f.Field = OXM_FIELD_IPV6_DST
+	f.HasMask = false
+
+	ipDstField := new(Ipv6DstField)
+	ipDstField.Ipv6Dst = ipDst
+	f.Value = ipDstField
+	f.Length = uint8(ipDstField.Len())
+
+	// Add the mask
+	if ipDstMask != nil {
+		mask := new(Ipv6DstField)
+		mask.Ipv6Dst = *ipDstMask
+		f.Mask = mask
+		f.HasMask = true
+		f.Length += uint8(mask.Len())
+	}
+
+	return f
+}
+
+// IP_ECN field
+type IpEcnField struct {
+	IpEcn uint8
+}
+
+func (m *IpEcnField) Len() uint16 {
+	return 1
+}
+func (m *IpEcnField) MarshalBinary() (data []byte, err error) {
+	data = make([]byte, 1)
+
+	data[0] = m.IpEcn
+	return
+}
+func (m *IpEcnField) UnmarshalBinary(data []byte) (err error) {
+	m.IpEcn = data[0]
+	return
+}
+
+// Return a MatchField for vlan id matching
+func NewIpEcnField(vlanPcp uint8) *MatchField {
+	f := new(MatchField)
+	f.Class = OXM_CLASS_OPENFLOW_BASIC
+	f.Field = OXM_FIELD_IP_ECN
+	f.HasMask = false
+
+	vlanPcpField := new(IpEcnField)
+	f.Value = vlanPcpField
+	f.Length = uint8(vlanPcpField.Len())
+
+	return f
+}
+
+// IP_PROTO field
+type IpProtoField struct {
+	protocol uint8
+}
+
+func (m *IpProtoField) Len() uint16 {
+	return 1
+}
+func (m *IpProtoField) MarshalBinary() (data []byte, err error) {
+	data = make([]byte, 1)
+	data[0] = m.protocol
+	return
+}
+
+func (m *IpProtoField) UnmarshalBinary(data []byte) error {
+	m.protocol = data[0]
+	return nil
+}
+
+// Return a MatchField for ipv4 protocol
+func NewIpProtoField(protocol uint8) *MatchField {
+	f := new(MatchField)
+	f.Class = OXM_CLASS_OPENFLOW_BASIC
+	f.Field = OXM_FIELD_IP_PROTO
+	f.HasMask = false
+
+	ipProtoField := new(IpProtoField)
+	ipProtoField.protocol = protocol
+	f.Value = ipProtoField
+	f.Length = uint8(ipProtoField.Len())
+
+	return f
+}
+
+// IP_DSCP field
+type IpDscpField struct {
+	dscp uint8
```

Comment:
Hi @ashish-varma , would you change this field as public? I know it is not introduced in your patch, but we need to set a masked field for IpDscpField, but the existing struct doesn't support this.

Reply:
Done.